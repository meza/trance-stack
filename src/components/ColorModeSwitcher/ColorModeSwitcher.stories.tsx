import React, { useContext, useEffect } from 'react';
import ColorModeSwitcher, { ColorMode, ColorModeContext } from './index';
import type { Meta, StoryFn, StoryObj } from '@storybook/react';

const ColorDecorator = (Story: StoryFn) => {
  const { colorMode } = useContext(ColorModeContext);

  const [cmode, setMode] = React.useState<ColorMode>(colorMode || ColorMode.LIGHT);

  useEffect(() => {
    document.firstElementChild?.classList.remove(ColorMode.LIGHT, ColorMode.DARK);
    document.firstElementChild?.classList.add(cmode);
  }, [cmode]);

  useEffect(() => {
    setMode(colorMode);
  }, [colorMode]);

  return (<div style={{ zoom: 5, display: 'grid', placeContent: 'center' }}>
    <ColorModeContext.Provider value={{ colorMode: cmode, setColorMode: setMode }}>
      <Story/>
    </ColorModeContext.Provider>
  </div>);
};

const meta = {
  component: ColorModeSwitcher,
  decorators: [ColorDecorator]
} satisfies Meta<typeof ColorModeSwitcher>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Toggle: Story = {};
