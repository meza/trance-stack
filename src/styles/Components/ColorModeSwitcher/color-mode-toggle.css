@import url("./sun-and-moon.css");

.color-mode-toggle-container {
  --size: 2rem;

  display: inline-block;
  padding: 3px;
}
.color-mode-toggle {
  --icon-fill: hsl(210deg 10% 30%);
  --icon-fill-hover: hsl(210deg 10% 15%);

  aspect-ratio: 1;
  background: none;
  block-size: var(--size);
  border: none;
  border-radius: 50%;

  /* Acknowledged issue but still nice to have on browsers that support it */

  /* doiuse-disable pointer */
  cursor: pointer;
  inline-size: var(--size);
  margin: 0;
  outline-offset: 5px;
  padding: 0;
  -webkit-tap-highlight-color: transparent;

  /* Acknowledged issue but still nice to have on browsers that support it */

  /* doiuse-disable css-touch-action */
  touch-action: manipulation;

  & > svg {
    block-size: 100%;
    inline-size: 100%;
    stroke-linecap: round;
  }

  @nest .dark & {
    --icon-fill: hsl(210deg 10% 70%);
    --icon-fill-hover: hsl(210deg 15% 90%);
  }

  @media (hover: none) {
    --size: 48px;
  }
}
