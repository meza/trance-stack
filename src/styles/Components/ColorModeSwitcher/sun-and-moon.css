.sun-and-moon {
  --ease-3: cubic-bezier(.25, 0, .3, 1);
  --ease-elastic-3: cubic-bezier(.5, 1.25, .75, 1.25);
  --ease-elastic-4: cubic-bezier(.5, 1.5, .75, 1.25);
  --ease-out-5: cubic-bezier(0, 0, 0, 1);

  & > :is(.moon, .sun, .sun-beams) {
    transform-origin: center center;
  }

  & > :is(.moon, .sun) {
    fill: var(--icon-fill);

    @nest .color-mode-toggle:is(:hover, :focus-visible) & {
      fill: var(--icon-fill-hover);
    }
  }

  & > .sun-beams {
    stroke: var(--icon-fill);
    stroke-width: 2px;

    @nest .color-mode-toggle:is(:hover, :focus-visible) & {
      stroke: var(--icon-fill-hover);
    }
  }

  @nest .dark & {
    & > .sun {
      transform: scale(1.75);
    }

    & > .sun-beams {
      opacity: 0;
    }

    & > .moon > circle {
      transform: translateX(-7px);

      @supports (cx: 1) {
        cx: 17;
        transform: translateX(0);
      }
    }
  }

  @media (--motion-ok) {
    & > .sun {
      transition: transform .5s var(--ease-elastic-3);
    }

    & > .sun-beams {
      transition: transform .5s var(--ease-elastic-4),
      opacity .5s var(--ease-3);
    }

    & .moon > circle {
      transition: transform .25s var(--ease-out-5);

      @supports (cx: 1) {
        transition: cx .25s var(--ease-out-5);
      }
    }

    @nest .dark & {
      & > .sun {
        transform: scale(1.75);
        transition-duration: .25s;
        transition-timing-function: var(--ease-3);
      }

      & > .sun-beams {
        transform: rotateZ(-25deg);
        transition-duration: .15s;
      }

      & > .moon > circle {
        transition-delay: .25s;
        transition-duration: .5s;
      }
    }
  }
}
