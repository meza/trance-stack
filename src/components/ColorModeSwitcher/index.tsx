import React, { useContext, useEffect } from 'react';
import { Form } from '@remix-run/react';
import { useTranslation } from 'react-i18next';
import { sensorScript, updateScript } from '~/components/ColorModeSwitcher/clientUtils';

export enum ColorMode {
  LIGHT = 'light',
  DARK = 'dark'
}

export interface ColorModeContextProps {
  colorMode: ColorMode;
  setColorMode: (colorMode: ColorMode) => void;
}

export const ColorModeContext = React.createContext<ColorModeContextProps>({} as ColorModeContextProps);

export interface ColorModeSensorProps {
  nonce: string | undefined;
}

export const ColorModeSensor = (props: ColorModeSensorProps) => {
  const inputProps: { nonce?: string } = {};
  if (props.nonce) {
    inputProps.nonce = props.nonce;
  }

  // detect the user's preferences and update the root element if it is not set already
  return (
    <>
      <script
        nonce={props.nonce}
        id={'color-mode-sensor'}
        suppressHydrationWarning
        dangerouslySetInnerHTML={{
          __html: `(${sensorScript.toString()})();`
        }}
      />
      <script
        nonce={props.nonce}
        suppressHydrationWarning
        id={'color-mode-update'}
        dangerouslySetInnerHTML={{
          __html: `(${updateScript.toString()})();`
        }}
      />
    </>
  );
};

/**
 `import { ColorModeSwitcher } from '~/components/ColorModeSwitcher';`

 Adds the ability to toggle between light and dark mode.

 ## Usage

 ```jsx
 <ColorModeSwitcher/>
 ```

 ## Provider

 In order to make it actually work, it needs to be wrapped in a `ColorModeContext.Provider` with the `colorMode` and `setColorMode` props set.

 ```jsx
 import { ColorMode, ColorModeContext } from '~/components/ColorModeSwitcher';
 import { useState } from 'react';

 const [colorMode, setColorMode] = useState<ColorMode>(ColorMode.LIGHT);

 <ColorModeContext.Provider value={{ colorMode: colorMode, setColorMode: setColorMode}}>
   <ColorModeSwitcher/>
 </ColorModeContext.Provider>
 ```

 **/
export default function ColorModeSwitcher() {
  const { colorMode, setColorMode } = useContext(ColorModeContext);
  // defaulting to the ColorMode.LIGHT makes react not moan about uncontrolled components as the hidden input needs a mode
  // the colorMode would be null if it's not set by the server
  const [nextColorMode, setNextColorMode] = React.useState<ColorMode>(colorMode === ColorMode.LIGHT ? ColorMode.DARK : ColorMode.LIGHT);
  const { t } = useTranslation();

  useEffect(() => {
    setNextColorMode(colorMode === ColorMode.LIGHT ? ColorMode.DARK : ColorMode.LIGHT);
  }, [colorMode]);

  useEffect(() => {
    // if there's no colour preference set, update the button in relation to the html tag
    if (!colorMode) {
      const container = document.firstElementChild;
      const cl = container?.classList;
      if (cl?.contains(ColorMode.DARK)) {
        setNextColorMode(ColorMode.LIGHT);
      } else {
        setNextColorMode(ColorMode.DARK);
      }
    }
  }, [colorMode]);

  const toggleColorMode = () => {
    setColorMode(nextColorMode);
  };

  return (
    <div className={'color-mode-toggle-container'}>
      {/*Awesome switcher design by @AdamArgyleInk: https://youtu.be/kZiS1QStIWc*/}
      <Form method='post' action='/settings/color-mode' replace>
        <input type={'hidden'} name={'colorMode'} value={nextColorMode}/>
        <button
          type={'submit'}
          className='color-mode-toggle'
          id='colorMode-toggle'
          title={t('translation:app.colorModeToggle', { nextMode: nextColorMode }).toString()}
          aria-live='polite'
          onClick={toggleColorMode}>
          <svg className='sun-and-moon' aria-hidden='true' width='24' height='24' viewBox='0 0 24 24'>
            <mask className='moon' id='moon-mask'>
              <rect x='0' y='0' width='100%' height='100%' fill='white'/>
              <circle cx='24' cy='10' r='6' fill='black'/>
            </mask>
            <circle className='sun' cx='12' cy='12' r='6' mask='url(#moon-mask)' fill='currentColor'/>
            <g className='sun-beams' stroke='currentColor'>
              <line x1='12' y1='1' x2='12' y2='3'/>
              <line x1='12' y1='21' x2='12' y2='23'/>
              <line x1='4.22' y1='4.22' x2='5.64' y2='5.64'/>
              <line x1='18.36' y1='18.36' x2='19.78' y2='19.78'/>
              <line x1='1' y1='12' x2='3' y2='12'/>
              <line x1='21' y1='12' x2='23' y2='12'/>
              <line x1='4.22' y1='19.78' x2='5.64' y2='18.36'/>
              <line x1='18.36' y1='5.64' x2='19.78' y2='4.22'/>
            </g>
          </svg>
        </button>
      </Form>
    </div>
  );
}
