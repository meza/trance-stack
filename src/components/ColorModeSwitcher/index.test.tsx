import { Form } from '@remix-run/react';
import { fireEvent, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { renderWithi18n } from '@test';
import ColorModeSwitcher, { ColorMode, ColorModeContext, ColorModeSensor } from './index';
import type { FormProps } from '@remix-run/react';
import type { UserEvent } from '@testing-library/user-event/setup/setup';
import type { FormEvent, PropsWithChildren, ReactElement } from 'react';

interface LocalTestContext {
  user: UserEvent;
}

vi.mock('@remix-run/react', async () => {
  const actual = (await vi.importActual('@remix-run/react')) as object;
  return {
    ...actual,
    Form: vi.fn()
  };
});

describe('The Color Mode Switch', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  describe('The color mode switcher', () => {
    beforeEach<LocalTestContext>((context) => {
      context.user = userEvent.setup();

      vi.mocked(Form).mockImplementation(({ children, ...props }: PropsWithChildren<FormProps>) => {
        const newProps = {
          ...props,
          onSubmit: (e: FormEvent) => {
            e.preventDefault();
          }
        };
        delete newProps.replace; // have to take this out, thanks dom.

        expect(props).toContain({ replace: true });

        return <form {...newProps}>{children}</form>;
      });
    });

    it('should configure the form correctly', () => {
      const component = renderWithi18n(<ColorModeSwitcher/>);
      expect(component.asFragment()).toMatchSnapshot();
    });

    it('should use the correct mode based on the html tag', () => {
      const ColorWrapper = ({ children, color }: PropsWithChildren<{ color: string }>) => {
        document.firstElementChild?.classList.add(color);
        return <>{children}</>;
      };

      const themeRender = (ui: ReactElement, color: 'dark' | 'light') => {
        return renderWithi18n(ui, {
          wrapper: ({ children }) => <ColorWrapper color={color}>{children}</ColorWrapper>
        });
      };

      const lightComponent = themeRender(<ColorModeSwitcher/>, 'light');
      const inputLight = lightComponent.container.querySelector('input[name="colorMode"]') as HTMLInputElement;
      expect(inputLight.value).toEqual(ColorMode.DARK);

      const darkComponent = themeRender(<ColorModeSwitcher/>, 'dark');
      const inputDark = darkComponent.container.querySelector('input[name="colorMode"]') as HTMLInputElement;
      expect(inputDark.value).toEqual(ColorMode.LIGHT);
    });

    it<LocalTestContext>('should change the color mode', async ({ user }) => {
      let colorMode = ColorMode.LIGHT;
      const setColorModeCallback = vi.fn().mockImplementation((newColorMode: ColorMode) => {
        colorMode = newColorMode;
        return newColorMode;
      });
      const component = renderWithi18n(
        <ColorModeContext.Provider value={{ colorMode: colorMode, setColorMode: setColorModeCallback }}>
          <ColorModeSwitcher/>
        </ColorModeContext.Provider>
      );

      const input = component.container.querySelector('input[name="colorMode"]') as HTMLInputElement;
      expect(input.value).toEqual(ColorMode.DARK);
      fireEvent.click(component.getByRole('button'));

      component.rerender(<ColorModeContext.Provider value={{ colorMode: colorMode, setColorMode: setColorModeCallback }}>
        <ColorModeSwitcher/>
      </ColorModeContext.Provider>);

      const input2 = component.container.querySelector('input[name="colorMode"]') as HTMLInputElement;
      expect(input2.value).toEqual(ColorMode.LIGHT);

      fireEvent.click(component.getByRole('button'));

      component.rerender(<ColorModeContext.Provider value={{ colorMode: colorMode, setColorMode: setColorModeCallback }}>
        <ColorModeSwitcher/>
      </ColorModeContext.Provider>);

      const input3 = component.container.querySelector('input[name="colorMode"]') as HTMLInputElement;
      expect(input3.value).toEqual(ColorMode.DARK);

      expect(setColorModeCallback).toHaveBeenCalledTimes(2);
      expect(setColorModeCallback).toHaveBeenNthCalledWith(1, ColorMode.DARK);
      expect(setColorModeCallback).toHaveBeenNthCalledWith(2, ColorMode.LIGHT);

    });
  });

  describe('The color mode sensor', () => {
    it('should render the correct script', () => {
      const nonce = 'example-nonce';
      const component = render(<ColorModeSensor nonce={nonce}/>);
      expect(component.asFragment()).toMatchSnapshot();
    });
  });
});
