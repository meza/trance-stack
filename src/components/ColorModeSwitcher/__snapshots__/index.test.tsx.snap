// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The Color Mode Switch > The color mode sensor > should render the correct script 1`] = `
<DocumentFragment>
  <script
    id="color-mode-sensor"
    nonce="example-nonce"
  >
    (() =&gt; {
  const cl = document.firstElementChild?.classList;
  if (!(cl?.contains("dark") || cl?.contains("light"))) {
    cl?.add(window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light");
  }
})();
  </script>
  <script
    id="color-mode-update"
    nonce="example-nonce"
  >
    (() =&gt; {
  addEventListener("load", () =&gt; {
    window.matchMedia("(prefers-color-scheme: dark)").addEventListener("change", (e) =&gt; {
      const cl = document.firstElementChild?.classList;
      cl?.remove(e.matches ? "light" : "dark");
      cl?.add(e.matches ? "dark" : "light");
    });
  });
})();
  </script>
</DocumentFragment>
`;

exports[`The Color Mode Switch > The color mode switcher > should configure the form correctly 1`] = `
<DocumentFragment>
  <div
    class="color-mode-toggle-container"
  >
    <form
      action="/settings/color-mode"
      method="post"
    >
      <input
        name="colorMode"
        type="hidden"
        value="dark"
      />
      <button
        aria-live="polite"
        class="color-mode-toggle"
        id="colorMode-toggle"
        title="app.colorModeToggle"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="sun-and-moon"
          height="24"
          viewBox="0 0 24 24"
          width="24"
        >
          <mask
            class="moon"
            id="moon-mask"
          >
            <rect
              fill="white"
              height="100%"
              width="100%"
              x="0"
              y="0"
            />
            <circle
              cx="24"
              cy="10"
              fill="black"
              r="6"
            />
          </mask>
          <circle
            class="sun"
            cx="12"
            cy="12"
            fill="currentColor"
            mask="url(#moon-mask)"
            r="6"
          />
          <g
            class="sun-beams"
            stroke="currentColor"
          >
            <line
              x1="12"
              x2="12"
              y1="1"
              y2="3"
            />
            <line
              x1="12"
              x2="12"
              y1="21"
              y2="23"
            />
            <line
              x1="4.22"
              x2="5.64"
              y1="4.22"
              y2="5.64"
            />
            <line
              x1="18.36"
              x2="19.78"
              y1="18.36"
              y2="19.78"
            />
            <line
              x1="1"
              x2="3"
              y1="12"
              y2="12"
            />
            <line
              x1="21"
              x2="23"
              y1="12"
              y2="12"
            />
            <line
              x1="4.22"
              x2="5.64"
              y1="19.78"
              y2="18.36"
            />
            <line
              x1="18.36"
              x2="19.78"
              y1="5.64"
              y2="4.22"
            />
          </g>
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
